### ISSUE TITLE
Typed Maps - like interfaces in TypeScript

### ISSUE THREAD

**Issue by KoTTi97** [dart-lang/language#783]
I would love to see typed Maps in Dart like it is possible in TypeScript.

TypeScript example:

```TypeScript
interface IPerson {
    name: string;
    age: number;
    height?: number;
}

const person: IPerson = {
    name: "Max",
    age: 26
}
```

I wish Dart would support something like that so i do not have to create a class for it. It could look something like this:

```Dart
interface IPerson {
  String name;
  int age;
  height? double;
}

Map<IPerson> person = {
  name: "Max",
  age: 26,
} 
```

**Comment by vanesyan** [comment]
You can already do it by using `abstract class`.
```dart
abstract class A {
  String get name;
  int get age;
  double? get height;
}

class B implements A {
  B({this.name, this.age, this.height});

  final int age;
  final String name;
  final double? height;
}
```

Still it is not possible to make map to implement abstract class as `Map` is just yet another class and thus `{ key: value }` is just consice instantiation of that class.

**Comment by Cat-sushi** [comment]
@KoTTi97 
If you want to construct instances in map like syntax without manually defining such constructor, then #698 might be a solution.

**Comment by rrousselGit** [comment]
@vanesyan that's a different thing

Typed maps, aka structures/records, have a very different assignment behavior.

**Comment by MrCreeper1008** [comment]
The "map" in TypeScript is different because it's a class, not a map. There's a `Map` class in TypeScript/JavaScript that is basically the same as maps in Dart. Plus, it's already possible to do what you want in Dart:
```dart
class Person {
  final String name;
  final int age;
  final double height;

  const Person({
    this.name,
    this.age,
    this.height,
  });
}

final Person person = Person(
  name: 'Max',
  age: 26
);
```

Introducing this basically means mixing two completely different concepts.

**Comment by g5becks** [comment]
@KoTTi97  it's not maps you want, in typescript / javascript, these are called object literals. I would love to see this represented in dart as well via anonymous classes, the main use case for me is simpler json encodeing/decoding. In Go they are quite convenient as well as they remove a lot of boilerplate code that is created soley for sending / recieving responses.

```go

car := struct {
	make    string
	model   string
	mileage int
}{
	make:    "Ford",
	model:   "Taurus",
	mileage: 200000,
}
```

Whats missing in dart is a way to declare a type inline as opposed to using a class declaration. The typescript example you showed doesn't really buy you anything when compared to what dart can already do, when this becomes useful is when declaring generic types, function return types and parameters, also variable types 
```typescript
const someFunc = (user: {name: string, age: number}): {data: {didSucceed: boolean } } => {
    const data: { didSucceed: boolean } = methodCall()
    return {data}
}
```
In dart you would have to define a class for every type that was defined inline here. 

It would be nice to have something similar in dart, but - It will probably never happen - Typescript and Go are structurally typed, 
and Dart isn't.  [C# has them](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/anonymous-types), but they suffer from the same problems that dart does due to the type system and  are only useful in the context of linq, they can't be used anywhere else. 

I think Data classes are a great alternative and will alleviate most if not all of the pain in situations where anonymous objects would be used.

**Comment by lrhn** [comment]
This looks something like Java's anonymous classes. In Dart, that would probably look like:
```dart
var foo = new Object() {
  int something = 42;
  String message = "Banana";
};
```
It won't give you any *type* for it, you'll still have to declare an interface:
```dart
abstract class Foo {
  abstract final int something;
  abstract final String message;
}
...
  var foo = const Foo() {
    final int something = 42; 
    final String message = "some text";
  };
```
Then `foo` will have a type which implements `Foo`.

**Comment by rrousselGit** [comment]
Anonymous classes aren't the same thing imo.
Anonymous classes is about first-class citizen class. 

Typed maps/structures/records are a mechanism separate from classes. They don't have constructors/inheritance.

**Comment by g5becks** [comment]
@lrhn 

I'm not sure that buys you anything over data classes for simple objects that are just containers for a set of fields with different types.

It might be useful if you are actually implementing an interface's methods though. 

```dart
abstract class Foo {
  Future<bool> doSomething();
}

var foo = const Foo() {
   Future<bool> doSomething() async {
      return false;
 }
}
```
I can't really think of a use case for this off the top of my head, but [F# has them](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/object-expressions) so I guess they are a useful somehow?

Seeing as how Dart has a nominal type system, I'm not sure how anonymous objects with structural equality could be supported (aside from some from of runtime reflection maybe? ), but for the op - If/when Dart lands support for [Type Aliases](https://github.com/dart-lang/language/issues/65) and [Sum Types](https://github.com/dart-lang/language/issues/546) , you can (almost) solve the same set of problems in a different way, E.G.

```dart
enum SomeUnionType {
    OneType,
    AnotherType,
    LastType,
}
typedef MyType = Map<String, SomeUnionType>;

MyType  myInstance =  { "key1": OneType(), "key2": AnotherType(), "key3": LastType() };

// destructuring based on key name would help also if this feature is added. 
var { key1, key2, key3 } = myInstance; 

```

**Comment by lrhn** [comment]
Dart has structural types too: Function types, `FutureOr` and (with Null Safety) nullable types. It's not being non-nominal that's the biggest challenge, it's having a type which is not assignable to `Object?`. The structural types that we have are still backed by actual objects.

**Comment by g5becks** [comment]
I'm not a language designer, so I am not so sure about the implementation. In typescript, the type doesn't exist at runtime anyway so they can just do whatever. For Go, I'm not sure - the struct cant be assigned to interface{} unless it's a pointer. In Scala the type is assignable to Object, but then runtime reflection is used for field access, which obviously won't work for dart.

**Comment by jodinathan** [comment]
I miss this too.  
The point is that the interface should not exist so it would be basically a desugar/analyzer thing:

```dart

interface Foo {
  String bar;
}

// example 1
var baz = <String, dynamic>{
  'bar': 'hi'
};
var foo = baz as Foo;

print(foo.bar); // prints hi
// the above would be basically a de-sugar to print(foo['bar']);

// example 2
var baz = Foo()..bar = 'hello';
// desugering to
var baz = <String, dynamic>{
  'bar': 'hello'
};
```

**Comment by jodinathan** [comment]
I guess this _kinda_ exists when you use JS interop:

```dart
@JS()
@anonymous
class Foo {
  external String get bar;
  external set bar(String n);

  external factory Foo({String bar});
}

var foo = Foo()..bar = 'hi';
// I guess the above line transpiles to something like in JS:
let foo = {
  'bar': 'hi'
};
```

**Comment by cedvdb** [comment]
For decoding where you have to changing values a lot, putting them in a map where each key is a string makes more sens than to create a class for each return value.  The problem is that the compiler doesn't know which key the map contains. I think this is what people would like. That the compiler knows which key the map contains without having to explicitly tell the compiler (via a class or alternatively enums for the keys)

**Comment by lrhn** [comment]
If you can specify the name and type of each entry in the "map" (and you need to if you're going to have any kind of static typing), that sounds like all the information needed for declaring a class.
All you need is a more concise "data-class" syntax. If you could write, say:
```dart
class Person(String name, int age, [num? height]);
```
to declare that class, then using a map-like construct doesn't seem particularly enticing any more.

**Comment by kasperpeulen** [comment]
This proposal seems very much in line with how Typescript does it:
https://github.com/dart-lang/language/blob/master/working/0546-patterns/records-feature-specification.md

It also based on structural typing instead of nominal typing.
I guess the opening post would look like this with Records.

```dart
typedef Person = {String name, int age, num? height};

const Person person = (
  name: "Max",
  age: 26
);
```

**Comment by cedvdb** [comment]
> ```dart
> typedef Person = {String name, int age, num? height};
> 
> const Person person = (
>   name: "Max",
>   age: 26,
>   height: null,
> );
> ```

That's awesome. Bonus point because I assume it would be easily serializable with json.

**Comment by kasperpeulen** [comment]
Yes, exactly, I didn't see anything about json serializing in the proposal, but I guess it could technically be done, and could also be faster than it is now if done with records.

**Comment by jodinathan** [comment]
the Records proposal along with meta programming proposal would rock Dart world

**Comment by cedvdb** [comment]
> the Records proposal along with meta programming proposal would rock Dart world

```
const Person person = (
  name: "Max",
  age: 26,
  height: null,
);
```

The properties have to be accessible without string like so `person.name`. This is looking a bit like the data class struct request.

**Comment by jodinathan** [comment]
@kasperpeulen https://github.com/dart-lang/language/issues/1482

**Comment by venkatd** [comment]
Structural typing would make a huge difference for interoperability with external systems. For example, we are working with a GraphQL API which is structurally typed.

Some parts of our code work with a `BasicUser` which is just id, name, photo. Other parts may need a `UserWithProfile` which has id, name, photo, and several other fields. We would want `UserWithProfile` to be usable in place of `BasicUser` since it's a strict superset.

Because we effectively want structural typing here, we hack around it by explicitly implementing a bunch of interfaces in our generated code.

In TypeScript, this is implemented as a `Pick`:
https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys

**Comment by jodinathan** [comment]
is this on the radar?  
or another solution because working with json in dart is way too boring and very time consuming

**Comment by jakemac53** [comment]
https://github.com/dart-lang/language/blob/master/working/1426-extension-types/feature-specification-views.md I think this proposal essentially covers this use case? cc @eernstg

**Comment by jodinathan** [comment]
is this the issue in the language project funnel? https://github.com/dart-lang/language/issues/1474

**Comment by jakemac53** [comment]
> is this the issue in the language project funnel? #1474

Looks like it to me :)

**Comment by eernstg** [comment]
I think there are several topics in this issue. I'll say something about a couple of them.

Let's consider [the original example](https://github.com/dart-lang/language/issues/783#issue-551017998):

```dart
interface IPerson {
  String name;
  int age;
  height? double;
}

Map<IPerson> person = {
  name: "Max",
  age: 26,
}
```

We don't (yet) have interface declarations in Dart, but the would presumably be a special variant of classes that support `implements` (so clients can create subtypes of them), but not `extends` or `with`. So the interface declared here would presumably work in a similar way as

```dart
abstract class IPerson {
  abstract String name;
  abstract int age;
  abstract height? double;
  // And we could add a private constructor to prevent
  // `extends IPerson` in another library.
}
```

`Map<IPerson>` would have to be something like `ObjectMap<IPerson>`, because `Map` takes two type arguments (and it's not very interesting to drag the discussion about static overloading of type parameter lists into this discussion, so we might as well just use different names for now). `ObjectMap` would be a new platform class, with some special magic.

In particular, we'd use plain identifiers to denote each key, requiring that is the name of a member of `IPerson`, and the corresponding value would be statically checked to have a type which is assignable to the given member.

We could do all these things, but there are many missing pieces of the puzzle:

Should `ObjectMap<IPerson>` be a subtype of `IPerson`? Should `ObjectMap<IPerson>` be a subtype of `Map<K, V>` for any `K`, `V` (maybe `Symbol` and `dynamic`)? What do we do if `IPerson` and `Map` both declare a member with the same name and incompatible signatures?

We could also ask whether `ObjectMap<C>` would form a subtype graph which is covariant in `C`, e.g., whether `ObjectMap<int>` would be a subtype of `ObjectMap<num>`?

We could ask whether `C` in `ObjectMap<C>` would have to satisfy some extra constraints, e.g., that it is a compile-time error if `C` declares any static members, or any instance members which are not instance variables, or if those instance variables are `final` or `late`, etc.etc., such that the behavior of an instance of type `ObjectMap<C>` could work like a map and also like an instance of `C`.

We could ask whether there would be a structural subtype relationship, that is, `ObjectMap<C1>` would be a subtype of `ObjectMap<C2>` if the set of members of `C2` is a subset of the set of members of `C1` (and in that case we wouldn't require that there is _any_ subtype relationship between `C1` and `C2`).

My take on this is that we could go down this path, but it does involve a lot of detailed language design (and possibly a large amount of implementation work in order to handle new kinds of subtyping relationships), and it's not obvious to me that it is a good fit for the language: Dart objects are not maps, and it would be a huge change to make them so. (I think that's a feature, not a bug. ;-)

---

However, we could turn this around and consider a possible underlying feature request which is much more compatible with Dart. Let's reconsider the example [here](https://github.com/dart-lang/language/issues/783#issuecomment-578576484) and assume that we have [enhanced default constructors](https://github.com/dart-lang/language/issues/698):

```dart
class Person {
  final String name;
  final int age;
  final double? height;
}

final Person person = Person(
  name: 'Max',
  age: 26,
);
```

This doesn't involve anonymous types, or structural subtyping, but it does allow for the declaration of a class and construction of instances based on a syntax which is similarly concise as the original example, and it preserves full static typing in a completely straightforward manner.

---

If the point is, instead, that we want to treat certain maps safely as certain existing class types then it is indeed possible to use a [`view`](https://github.com/dart-lang/language/blob/master/working/1426-extension-types/feature-specification-views.md) to do it:

```dart
abstract class IPerson {
  abstract String name;
  abstract int age;
  abstract height? double;
}

view IPersonMap on Map<Symbol, Object?> implements IPerson {
  String get name => this[#name] as String;
  set name(String value) => this[#name] = value;
  // and similarly for `age` and `double`.
}

void main() {
  Map<Symbol, Object?> map = { #name: 'Max', #age: 26 };
  IPersonMap p = map;

  // `p` is now treated statically safely, with the same interface as an `IPerson`.
  print(p.name);
  p.name = 'Joe';
}
```

The view is a static mechanism (and not yet part of the language), but it includes a `box` getter that returns an instance whose dynamic (and static) type is `IPerson`, and that's a full-fledged object which would also support things like dynamic method invocations and `is` tests.

This is of course a lot more verbose, and the purpose is completely different: This mechanism is aimed at supporting a manually specified statically safe treatment of objects of a freely chosen underlying implementation type. The point is that the implementation type (`Map<Symbol, dynamic>` here) is unsafe, and we want to restrict the usage of the given map to an existing interface, and then we just have to implement the latter in terms of the former (in this case: translating object member accesses to map operations).

**Comment by jodinathan** [comment]
`views` is a very interesting feature @eernstg.  

I can see a builder to make it easy to JSON interop. Would the below work?

```dart
// we take an abstract class and add the Interface annotation
@Interface()
abstract class Person {
  abstract String name;
  abstract int age;
  abstract height? double;
}

// generate the view Interface through the builder
view PersonInterface on Map<String, Object?> implements Person {
  String get name => this['name'] as String;
  set name(String value) => this['name'] = value;

  // and similarly for `age` and `double`.
}

// declare some API bindings
class SomeApi {
  // expose the Person endpoint
  Future<Person> fetchPerson() async {
    // fetch some map
    final resultMap = await ajaxSomePersonMap();
   
    // return as the Person interface
    return resultMap as PersonInterface;
  }
}

Future<void> main() async {
  final api = SomeApi();
  // call the API in a object oriented way
  final person = await api.fetchPerson();

  // typed =]
  print(person.name);
}
```

This along with static meta programming can finally make working with json easier and satisfying.  

Question: I understood from the proposal that `views` are going to be very lightweight, thus our example should be pretty close to manually using a `Map`, right?

**Comment by eernstg** [comment]
@jodinathan, we'd need a couple of adjustments:

```dart
@Interface()
abstract class Person {...} // Same as before.

view PersonInterface on Map<String, Object?> implements Person {...} // Same as before.

class SomeApi {
  Future<Person> fetchPerson() async {
    final resultMap = await ajaxSomePersonMap();
    return resultMap.box;
  }
}

void someOtherFunction(Person p) {...}

Future<void> main() async {
  final api = SomeApi();
  final person = await api.fetchPerson();

  // Statically typed; `person` has type `Person`.
  print(person.name);
  someOtherFunction(person);
}
```

This would work, and the object which is the value of `person` in `main` would be a full-fledged Dart object of type `Person`. In this case you're paying (in terms of time and space) for a wrapper object, and in return you get normal object-oriented abstraction (you can assign `person` to a variable of type `Object` and restore the type by doing `is Person`) as well as dynamic invocations.

You could also maintain a more lightweight approach where the map isn't wrapped in a different object. In this case all instance method invocations would be static (that is, there is no OO dispatch and they could be inlined), but in return you must maintain the information that this is actually a `PersonInterface` view, and not an actual object of type `Person`.

But you still have the `Person` interface, and it is still checked statically that you only access the underlying map using those view method implementations:

```dart
@Interface()
abstract class Person {...} // Same as before.

view PersonInterface on Map<String, Object?> implements Person {...} // Same as before.

void someOtherFunction(Person p) {...}

class SomeApi {
  Future<PersonInterface> fetchPerson() async => await ajaxSomePersonMap();
}

Future<void> main() async {
  final api = SomeApi();
  final person = await api.fetchPerson();

  // Statically typed; `person` has type `PersonInterface`.
  print(person.name);

  // You can still use `person` as a `Person`, but then it must be boxed.
  someOtherFunction(person.box); // Just passing `person` is a compile-time error.
}
```

The point is that you can pass the `Map` around under the type `PersonInterface`, and this means that you can store it in a list or pass it from one function to another, and you'll never need to pay (time or space) for a wrapper object. But if you insist on forgetting that it is a view then you must box the map, which is then simply a normal Dart object of type `Person`.

**Comment by jodinathan** [comment]
in the json use case, the view is enough.  

can we pass the interface around?  

```dart
view PersonInterface on Map<String, Object?> implements Person {...} // Same as before.

void someOtherFunction(PersonInterface p) {
  print(p.name);
}

Future<void> main() async {
  final api = SomeApi();
  final person = await api.fetchPerson();

  someOtherFunction(person);
}
```

**Comment by eernstg** [comment]
Yes, inside `someOtherFunction` it is statically known that `p` is a `Map<String, Object?>` which must be handled using the members of `PersonInterface`, so there's no need for a wrapper object. On the other hand, then we can't invoke `someOtherFunction` on an argument of type `Person`, so `Person` and `PersonInterface` are strictly separate types, statically and dynamically. They just happen to support exactly the same members because we have constrained `PersonInterface` to implement `Person` (OK, `PersonInterface` could have additional members as well, but it can't be incomplete or contradict anything in `Person`).

**Comment by jodinathan** [comment]
@eernstg this is awesome!  

We really need this 🥺

**Comment by cedvdb** [comment]
@eernstg could the view be arbitrary partial ?  The idea is if I have a class `A` with 100 properties and want a `Partial<A>` view that has any combination of the properties of `<A>`. This is an extremely useful thing to have, the use case is for the U in CRUD, when you can update an arbitrary property of a Database record.

**Comment by ciriousjoker** [comment]
Any progress? Typescript is amazing in this regard.

**Comment by cedvdb** [comment]
The lack of this feature is error prone when adding a property to an serializable class. The fromMap will fail to compile while the toMap won't. Hopefully macros will help there.

**Comment by bernaferrari** [comment]
Records feature is kind of this.

**Comment by ddzy** [comment]
any progress?

**Comment by bernaferrari** [comment]
This is done already..

**Comment by lrhn** [comment]
It is very, very unlikely that Dart will introduce a "map type" which you can check whether an existing normal map "implements".
Can't do `someMap is {"name": String, "age": int}` and then somehow know that a later `someMap["name"]` has type `String`, because Dart maps cannot guarantee that they preserve map type or value over time.
Instead you should check the structure and extract the values at the same time, which is exactly what map patterns do.

For having simple structures with typed fields, Records is probably the most direct implementation of what's being asked for here, and they exist since Dart 3.0.
Except for the optional value, where you would have to explicitly write the `null`.

The next best thing would be "data classes", which could be defined something like my example above.

The one thing that's not being asked for, but is probably wanted, is JSON encoding and decoding. 
Records do not support that. Data classes could using macros.

The value will not _be_ a `Map`. The type system is unlikely to get a special map type which knows the type of its individual entry values (and which only allows string keys). It's not a good fit for a language which, well, is not JavaScript
All JavaScript objects are "maps", so class types are mostly "map types".
In other languages, maps are homogenous data structures, and objects implement interfaces, two completely different kinds of objects.

Still, it's not completely _impossible_.
Dart could introduce a "typed map" type, where the type carries the keys (which will likely need to be consistent and have primitive equality), and the instance carries the values. Very similar to records, a fixed structure parameterized by values. Using the `[]` operator on that type is not a method call if the argument is a constant, but a special syntactic form which has the type of the entry if the value is a constant. If not constant, or not called at that type, it is a method lookup which has type `Object?`, out maybe a specified supertype of all values. Typed maps can be immutable or you can be allowed to update the fields that exist, but not change the structure. Some fields can be optional, and their presence can be checked using `containsKey`. Can be assigned to `Map<Object?>`, but not to a typed map with a different structure.
Will need some way to build, likely a constructor. If the keys are all strings containing identifiers, then the construction makes can be inferred from that, otherwise they'll have to be positional.


It's just not particularly useful.
It's not a type that you can place on an existing map, to see if it fits, it's a new kind of object which plain maps do not have. It's just another class which implements `Map`.
JSON encoding won't preserve the knowledge that it comes from a typed map, so decoding will have to recreate it anyway. 
If you can call `[constant]`, you can do `.name`, so a data class which a macro makes implement `Map<String, Object?>`
would be just as powerful.

I don't see anything left to do here, which is at all likely, except possibly simple data classes.

**Comment by TekExplorer** [comment]
> It is very, very unlikely that Dart will introduce a "map type" which you can check whether an existing normal map "implements". Can't do `someMap is {"name": String, "age": int}` and then somehow know that a later `someMap["name"]` has type `String`, because Dart maps cannot guarantee that they preserve map type or value over time. Instead you should check the structure and extract the values at the same time, which is exactly what map patterns do.


see pattern matching
```dart
if (map case {'name': String name, 'age': int age}) {
   print('name: $name, age: $age');
}
```

**Comment by TekExplorer** [comment]
also, records could support json encoding with macros too.

**Comment by alessandro-amos** [comment]
I didn't quite understand the problem ... isn't that what it's already saying?
```dart
typedef TPerson = ({ String name, int age });
TPerson person = (name: "teste", age: 10);
```

**Comment by bernaferrari** [comment]
No, because you lack person.values, keys, serialization and any functionality that expects a map

**Comment by TekExplorer** [comment]
Perhaps you want a map extension type?
```dart
extension type YourType(Map<String, dynamic> map) {
  String get thing => map['thing'] as String;
  int get i => map['i'] as int;
}
```

**Comment by cedvdb** [comment]
Strings are not safe

**Comment by bernaferrari** [comment]
Google (as well as many others) is using Zod to type a scheme for Gemini, so the model knows what it can return. When value is returned, it is already cast in the schema you provided. That's basically impossible in Dart right now, but this issue would allow it.

**Comment by lrhn** [comment]
Sounds like you'd want a Zod-to-object conversion. Or Zod-to-record.

Typed maps are an oxymoron. If it's a Dart `Map`, then it has precisely one key type and one value type. Having different types for different names is what objects are for.

It "works" in JavaScript/TypeScript because the type system isn't sound or safe. If you want unsound types in Dart, you need to add runtime type checks. The map itself cannot do that, so you need something around the map. Maybe an extension type.

**Comment by mavel124** [comment]
I support this question. Since Dart maps are used everywhere (if not by you, then in some library), it would be very helpful to have some tool for map typing.

**Comment by lrhn** [comment]
A tool for map typing is not necessarily the same as typed maps.

A type map would be a type that is a subtype of `Map` where the type of `map[someString]` depends on the _value_ of `someString`.
That's _dependent typing_ which is not easy to add to a sound type system.
(Typescript of not sound.)

Rather than introduce an unsound type like `{"foo": int,"bar": String?}` and try to apply that to maps, which can only ever affect the static type system when the string value is known at compile time, maybe introduce an extension type like: 
```dart
extension type FooBar(Map<String, Object?> _)
    implements Map<String, Object?> {
  int get foo=>_["foo"] as int;
  set foo(int v) { _["foo"]=v;}
  String? get bar =>_["bar"] as String?;
  set bar(String? v) { _["bar"]=v;}
}
```
It's more writing, but I can make your a small script to generate this extension type from the type map above, if that's a problem.

You can still look up anything using `[]`.

If you want to _enforce_ a type scheme for a map, you can make a wrapper Aarons the map that checks the values coming in, and throws if they're invalid.

But you can't have dependent typing in Dart. Not any time soon, at least. That's a very big mouthful, and unlikely to be worth the very significant effort to make it sound. (If that's even possible.)

**Comment by tatumizer** [comment]
Rather than defining an extension, you can define a map-backed class. I played with this idea years ago - see https://github.com/tatumizer/pigeon_map

As a bonus, you get a much more efficient implementation (both in terms of performance and memory footprint) of Map. There are other advantages (e.g. the ease of serialization).
The idea failed to attract much interest. :-(

**Comment by cyberalt1** [comment]
Open issue, 66 upvotes, 0 fix
